ably/realtime_presence_test.go

```diff
@@ -96,3 +104,38 @@ func TestRealtimePresence_Sync250(t *testing.T) {
 		t.Fatalf("members3: %v", err)
 	}
 }
+
+var stateGuard = ^ably.StateEnum(0)
+
+var presTransitions = []ably.StateEnum{
+	ably.StateConnConnecting,
+	ably.StateConnConnected,
+	stateGuard,
+	ably.StateChanAttaching,
+	ably.StateChanAttached,
+}
+
+func TestRealtimePresence_EnsureChannelIsAttached(t *testing.T) {
+	rec := ably.NewStateRecorder()
+	opts := rec.Options()
+	opts.NoConnect = true
+	app, client := testutil.ProvisionRealtime(nil, opts)
+	defer safeclose(t, client, app)
+	channel := client.Channels.Get("persisted:presence_fixtures")
+	if err := ably.Wait(client.Connection.Connect()); err != nil {
+		t.Fatal(err)
+	}
+	rec.Add(stateGuard)
```

Comment:
It's more subtle than that. I can't verify if rec.States() contains only [Connecting, Connected] since there may be states values buffered in the chan but not yet added to the slice. States() then will return [Connecting, Connected] and we'll be happy, but as soon as it returns the buffered in the channel states will then be added to the slice before `presence.Get` fires. What makes mark work is that it's sent via the same channel, that is it's enqueued at the end. Obviously it's not 100% accurate test, to be 100% sure we would need to introduce long timeouts before acknowledging the states, but in my opinion it's good enough for our case.

If we really don't want to enforce ordering that way, I surely can rewrite this as you suggested. However this way of testing is much easier to write more complicated tests which assert behaviour of the connection/channel due to various external factors (connection recovery), as it's as simple as compare all recorded events to single slice instead of comparing partial results each step (I'm actually using this more).

Besides the above I have no strong feelings for the mark or absence of it, can rewrite it if you feel it would make maintanence & development of ably-go easier :)